src/niri.rs

```diff
@@ -3728,47 +3734,160 @@ impl Niri {
+                            queue.pop().invalid_buffer();
+                            warn!("{err}");
+                        }
+                    }
+                };
+            }
+        }
+
+        self.screencopy_state = screencopy_state;
+    }
+
     pub fn render_for_screencopy(
         &mut self,
-        backend: &mut Backend,
+        renderer: &mut GlesRenderer,
+        manager: &ZwlrScreencopyManagerV1,
         screencopy: Screencopy,
     ) -> anyhow::Result<()> {
-        let output = screencopy.output().clone();
-        ensure!(self.output_state.contains_key(&output), "output is missing");
+        let _span = tracy_client::span!("Niri::render_for_screencopy");
 
-        self.layout.update_render_elements(&output);
+        let output = screencopy.output();
+        ensure!(self.output_state.contains_key(output), "output is missing");
 
-        backend
-            .with_primary_renderer(move |renderer| {
-                let elements = self
-                    .render(
-                        renderer,
-                        &output,
-                        screencopy.overlay_cursor(),
-                        RenderTarget::ScreenCapture,
-                    )
-                    .into_iter()
-                    .rev();
-
-                let region_loc = screencopy.region_loc();
-                let elements = elements.map(|element| {
-                    RelocateRenderElement::from_element(
-                        element,
-                        region_loc.upscale(-1),
-                        Relocate::Relative,
-                    )
-                });
+        self.layout.update_render_elements(output);
 
-                let scale = output.current_scale().fractional_scale().into();
-                let transform = output.current_transform();
-                render_to_shm(renderer, screencopy.buffer(), scale, transform, elements)
-                    .context("error rendering to screencopy shm buffer: {err:?}")?;
+        let elements = self.render(
+            renderer,
+            output,
+            screencopy.overlay_cursor(),
+            RenderTarget::ScreenCapture,
+        );
+        let queue = self
+            .screencopy_state
+            .get_mut(manager)
+            .context("error getting screencopy manager")?;
+        let damage_tracker = queue.split().0;
+
+        let render_result = Self::render_for_screencopy_internal(
+            renderer,
+            output,
+            &elements,
+            screencopy.with_damage(),
+            damage_tracker,
+            &screencopy,
+        )
+        .map(|_| ());
+
+        match render_result {
+            Ok(()) => screencopy.submit(false),
+            Err(_) => {
+                // recreate damage tracker to report full damage next check.
+                *damage_tracker = OutputDamageTracker::new((0, 0), 1.0, Transform::Normal);
+                screencopy.invalid_buffer();
```

Comment:
post_error() kills the screencopy client, so it should only be used for protocol violations. This should instead send the failed event.

Reply:
The docs say `invalid_buffer` is for when:
> buffer attributes are invalid

Isn't that when the format or size of the buffer is not the same as the ones sent by the manager? Those are checked in the `render_to_` functions, but I guess I need to separate protocol violation from other kinds of errors.

Reply:
I thought that this branch could trigger if the output resized between copy_with_damage() and rendering? Might be mistaken